### PR TITLE
feat(ui): start using Datatables to generate Objects table

### LINKED ISSUE #60: 2. Paginate objects result table inside job - ui

### LINKED ISSUE #80: 2. Switch to use `stats` field for `Job` and `JobBare` in `ui`
- `core` schema `Job` remove `objects` list from class.
- Update `ui` `model.py` to use supplied stats from `core`, implemented in #75.
- Populate stats for job in a projects job overview(`project.html`) site and detailed job view(`job.html`).

### PR DESCRIPTION
**Pull request by tomrtk**
Closes #60, closes #80.

### CODE REVIEW

**Comment by tomrtk** [comment]
Current state of table:
image

File: src/core/api/api.py [hunk 1]
```diff
@@ -611,3 +611,43 @@ async def get_object_image(
...
+
+@core_api.get(
+    "/projects/{project_id}/jobs/{job_id}/objects",
+    response_model=Dict[str, Any],
+)
+def get_objects_from_job(
+    project_id: int = Path(..., ge=1),
+    job_id: int = Path(..., ge=1),
+    start: int = Query(0, ge=0),
+    lenght: int = Query(10, ge=1),
```
Comment (R0flcopt3r) [inline src/core/api/api.py]:
```suggestion
    length: int = Query(10, ge=1),
```

File: src/core/api/api.py [hunk 2]
```diff
@@ -611,3 +611,43 @@ async def get_object_image(
...
+) -> Dict[str, Any]:
+    """Endpoint to get part of objects from a job.
+
+    Returns
+    -------
+    Dict[str, Any]
+        Containing total number of object in job and part of objects from
+        `start` to `start + lenght`.
```
Comment (R0flcopt3r) [inline src/core/api/api.py]:
```suggestion
        `start` to `start + length`.
```

File: src/core/api/api.py [hunk 3]
```diff
@@ -611,3 +611,43 @@ async def get_object_image(
...
+
+    Raises
+    ------
+    HTTPException
+        If no session to database can be established. Status code: 503.
+    HTTPException
+        If project or job id provided is not found. Status code: 422.
+    """
+    try:
+        data = services.get_job_objects(project_id, job_id, start, lenght)
```
Comment (R0flcopt3r) [inline src/core/api/api.py]:
```suggestion
        data = services.get_job_objects(project_id, job_id, start, length)
```

File: src/core/services.py [hunk 4]
```diff
@@ -508,3 +508,55 @@ def queue_job(project_id: int, job_id: int, session: Session) -> None:
...
+        return None
+
+    job = project.get_job(job_id)
+
+    if job is None:
+        return None
+
+    response: Dict[str, Any] = {}
+    response["total_objects"] = len(job._objects)
+    response["data"] = job._objects[start : start + lenght]
```
Comment (R0flcopt3r) [inline src/core/services.py]:
```suggestion
    response["data"] = job._objects[start : start + length]
```

File: src/core/services.py [hunk 5]
```diff
@@ -508,3 +508,55 @@ def queue_job(project_id: int, job_id: int, session: Session) -> None:
         logger.error(
             f"Cannot queue job {job_id}, it's of status {job.status()}."
         )
+
+
+def get_job_objects(
+    project_id: int, job_id: int, start: int, lenght: int
```
Comment (R0flcopt3r) [inline src/core/services.py]:
```suggestion
    project_id: int, job_id: int, start: int, length: int
```

File: src/ui/projects/api.py [hunk 6]
```diff
@@ -449,3 +449,37 @@ def check_api(self) -> bool:
             return False
 
         return True
+
+    def get_objects(
+        self, project_id: int, job_id: int, start: int, lenght: int
```
Comment (R0flcopt3r) [inline src/ui/projects/api.py]:
```suggestion
        self, project_id: int, job_id: int, start: int, length: int
```

File: src/ui/projects/utils.py [hunk 7]
```diff
@@ -0,0 +1,22 @@
+"""Utilities functions used in ui."""
+from pydantic import validate_arguments
+
+SQLITE_INTEGER_LIMIT = 2 ** 63 - 1
+
+
+@validate_arguments
+def validate_int(
```
Comment (R0flcopt3r) [inline src/ui/projects/utils.py:10]: Wouldn't it be better if it returned an `Optional[int]`?

File: src/ui/src/js/projects.js [hunk 8]
```diff
@@ -23,6 +23,8 @@ window.$ = window.jQuery = jQuery;
 
 import "jstree";
 
+require("datatables.net")(window, $);
```
Comment (R0flcopt3r) [inline src/ui/src/js/projects.js]: does this grab scripts from external sites?

File: src/core/api/api.py (hunk 2 above)
Comment (tomrtk) [reply src/core/api/api.py]: This PR is based upon #72 and this is from there. It is all ready fixed in that PR.

File: src/core/services.py (hunk 4 above)
Comment (tomrtk) [reply src/core/services.py]: This PR is based upon #72 and this is from there. It is all ready fixed in that PR.

File: src/ui/projects/api.py (hunk 6 above)
Comment (tomrtk) [reply src/ui/projects/api.py]: This PR is based upon #72 and this is from there. It is all ready fixed in that PR.

File: src/core/services.py (hunk 5 above)
Comment (tomrtk) [reply src/core/services.py]: This PR is based upon #72 and this is from there. It is all ready fixed in that PR.

File: src/ui/projects/utils.py (hunk 7 above)
Comment (tomrtk) [reply src/ui/projects/utils.py:8]: This PR is based upon #72 and this is from there. It is all ready fixed in that PR.

File: src/core/api/api.py (hunk 1 above)
Comment (tomrtk) [reply src/core/api/api.py]: This PR is based upon #72 and this is from there. It is all ready fixed in that PR.

File: src/ui/src/js/projects.js (hunk 8 above)
Comment (tomrtk) [reply src/ui/src/js/projects.js]: @MindTooth can probably answer this better. datatables is installed with npm and this indicate it is needed, similar to import, but the import syntax did not work for me. So I stayed with what worked, not using time to get it working.

Comment (MindTooth) [reply src/ui/src/js/projects.js]: This syntax is called CommonJS and is to be regardes as obsolete.  Now we use what’s called ESM.  `import` is newer and should be used.

Regarding `.net`, the package has it in its name. 😅

**Comment by tomrtk** [comment]
Remaining work:
- [x] ~~Use stats from core when #75 is merged. This can maybe be it's own issue and PR.~~ new issue #80 
- [x] Get preview popup to work correctly.

Feel free to work on popup if anyone has an idea of a good way of doing it.

And also please test run branch.

File: src/ui/src/js/projects.js [hunk 9]
```diff
@@ -23,6 +23,8 @@ window.$ = window.jQuery = jQuery;
 
 import "jstree";
 
+var dt = require("datatables.net")(window, $);
```
Comment (MindTooth) [inline src/ui/src/js/projects.js]:
```suggestion
import dt from "datatables.net";
dt(window, $);
```

So we use imports all over. 😉

**Review by Kad1n — COMMENTED** [review]
Seems to be some cases in `projects.py` that is not covered by tests. They are mostly error checks. Should perhaps write a quick test to cover them.

Otherwise this PR looks good :+1:

File: src/core/services.py [hunk 10]
```diff
@@ -563,7 +563,6 @@ def get_job_objects(
     response: Dict[str, Any] = {}
     response["total_objects"] = len(job._objects)
     response["data"] = job._objects[start : start + length]
-    print(response["data"])
```
Comment (R0flcopt3r) [inline src/core/services.py:566]: This should probably be removed?

File: src/ui/projects/api.py [hunk 11]
```diff
@@ -449,3 +447,39 @@ def check_api(self) -> bool:
...
+        response = self.get(
+            url,
+            params={"start": start, "length": length},
+        )
+
+        if isinstance(response, requests.Response):
+            dct = response.json()
+            objects = [Object(**d) for d in dct["data"]]
+            total_object_count = dct["total_objects"]
+            return objects, total_object_count
```
Comment (R0flcopt3r) [inline src/ui/projects/api.py:483]: Total amount of object should also be in `Job`/`JobBare` as part of `JobStat` if I remember correctly.

File: src/core/services.py (hunk 10 above)
Comment (tomrtk) [reply src/core/services.py:566]: ??
```diff
-    print(response["data"])
```

File: src/ui/projects/api.py (hunk 11 above)
Comment (tomrtk) [reply src/ui/projects/api.py:483]: The total amount is needed to be part of response to Datatable, so the object endpoint is returning it also.

File: src/core/services.py (hunk 10 above)
Comment (R0flcopt3r) [reply src/core/services.py:566]: brane go brrr.....

**Comment by tomrtk** [comment]
Added some more tests. Main contribution of PR should almost have 100% coverage. Some of the fixes and improvement in other parts of `ui` code base are partial testes since they do not have any tests from before.

File: src/ui/projects/api.py (hunk 11 above)
Comment (R0flcopt3r) [reply src/ui/projects/api.py:483]: I see!

File: src/ui/projects/api.py [hunk 12]
```diff
@@ -449,3 +447,39 @@ def check_api(self) -> bool:
             return False
 
         return True
+
+    def get_objects(
+        self, project_id: int, job_id: int, start: int, length: int
+    ) -> Optional[Tuple[List[Object], int]]:
+        """Get a list of Objects for `job_id` from `start` to `start + lenght`.
```
Comment (MindTooth) [inline src/ui/projects/api.py]:
```suggestion
        """Get a list of Objects for `job_id` from `start` to `start + length`.
```

File: src/ui/src/js/projects.js [hunk 13]
```diff
@@ -23,6 +23,9 @@ window.$ = window.jQuery = jQuery;
 
 import "jstree";
 
+import dt from "datatables.net";
```
Comment (MindTooth) [inline src/ui/src/js/projects.js:26]:
```suggestion
import 'datatables.net'
import 'datatables.net-dt'
```

Seemed they were both needed.  I'll test tomorrow.

File: tests/unit/ui/test_utils.py [hunk 14]
```diff
@@ -0,0 +1,20 @@
+"""Unit test of ui utils functions."""
+import pytest
+from pydantic.error_wrappers import ValidationError
+
+from ui.projects.utils import validate_int, SQLITE_INTEGER_LIMIT
+
+
+def test_vaidate_int() -> None:
```
Comment (MindTooth) [inline tests/unit/ui/test_utils.py]:
```suggestion
def test_validate_int() -> None:
```

**Review by MindTooth — APPROVED** [review]

**Pull request merged**